### DISCUSSION TITLE
Dynamic labels for builder blocks

### DISCUSSION THREAD

**Discussion by fayne1337** [filamentphp/filament#2961] [Help]
I'm creating a form using the builder component. Is there a way to make the block labels dynamic depending on a value inside this specific block? Each block has a child `TextInput` component named _label_ which I'd like to target.

I'm aware of the possibility to use the `$get()` function inside the label's closure, but since the blocks are repeatable and stored in an array (with hash keys), I'm not sure what parameter I could use to fetch the current block's child.

```
Components\Builder\Block::make('textfield')
   ->reactive()
   ->label(
      fn(Closure $get) => $get('.......')
   ),
```

**Comment by ryangjchandler** [comment]
I think you can already do this.. the `Block` class uses `HasLabel` which has a `label()` method like regular fields. The Blade view will also call `getLabel()` which tries to use the `HasLabel::getLabel()` method by default.

**Reply by fayne1337** [reply]
Yes, I'm aware of the `label()` method inherited from the `HasLabel` trait. I've tinkered with this specific method (see my code snippet in my request). I do not want to set a static label tho, but a value adapting dynamically to a sub fields' input value.

Usually I could just do
```
->label(fn(Closure $get) => $get('exmaple_field'))
```
but since the blocks are repeatable this does not work.

Here's a little more complete example:
```
return $form
      ->schema([
         Components\Builder::make('builder')
            ->schema([
               Components\Builder\Block::make('block')
                  ->reactive()
                  ->label(
                     fn(\Closure $get) => $get('label') // this one is not working
                  )
                  ->schema([
                     Components\TextInput::make('label')
                  ])
            ])
      ]);

```

**Reply by danharrin** [reply]
`$get('label')` will attempt to get another block' data with the array key "label".

`$state` will give you the content of that block in an array, so you need `$state['label']`

Also the label field needs to be `lazy()` so that the label updates on the block when you type

Also you need to check if $state is null and provide a fallback label for the block picker dropdown

**Reply by fayne1337** [reply]
> $state will give you the content of that block in an array, so you need $state['label']

I'm not sure this is correct, as $state seems to return an associative array containing **all blocks** and their data.

```
Array
(
    [bb26fbed-cb07-4e1a-9f9e-fa136be2336e] => Array
        (
            [type] => block
            [data] => Array
                (
                    [label] => 
                )

        )

    [774aadac-6e96-46e3-ae3a-1b6b6cc8916f] => Array
        (
            [type] => block
            [data] => Array
                (
                    [label] => 
                )

        )
)
```
so it seems I'd need to somehow get the current block's hash in order to do something like
```
->label(
   fn(array $state) => $state['{hash}']['data']['label'] ?? 'Fallback'
)
```

**Reply by danharrin** [reply]
Hmm `$state` might return all items when `label()` is called in the context of the block picker, but if its called when the block is associated with data, then it will return the correct block data.

**Reply by danharrin** [reply]
Okay I've just checked myself, you are correct

**Reply by danharrin** [reply]
Got it working

https://user-images.githubusercontent.com/41773797/176480999-b891bb29-1230-40f4-9eca-258848f17818.mov

**Reply by danharrin** [reply]
#2976

**Reply by fayne1337** [reply]
Awesome, that's exactly what I was trying to do. Thank you! :)
